### PR TITLE
Vberenz/normalize

### PR DESCRIPTION
**Pull request by vincentberenz**
[//]: # "Thanks for your contribution.  To make life of the reviewers easier,"
[//]: # "please give this pull request a meaningful title and provide the"
[//]: # "requested information below."

## Description

Hysr control returns information related to the system (simulation, robot, balls ...) in terms of instances of "State" (e.g. MainSimState, etc), but some may need instead numpy list of normalized floats (e.g. as observations in gym environment).

Code was added to "cast" instances of State to normalized numpy arrays of floats.

## How I Tested

- Added unit tests

## I fulfilled the following requirements

[//]: # "Please make sure you followed these steps before requesting a review."
[//]: # "Check the boxes in the list below, when done."

- [x] All new code is formatted according to our style guide (for C++ run clang-format, for Python, run flake8 and fix all warnings).
- [x] All new functions/classes are documented and existing documentation is updated according to changes.
- [x] No commented code from testing/debugging is kept (unless there is a good reason to keep it).

### CODE REVIEW

**Review by luator — COMMENTED** [review]
Looks good to me.  I have a few suggestions how the code could be simplified a bit but nothing critical, so feel free to ignore, if you don't agree.

**Review by luator — APPROVED** [review]

File: hysr/normalize.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+    return (0.0, 0.0, 0.0)
+
+
+def normalize_orientation3D(orientation: Orientation3D) -> Orientation3D:
+    """
+    Normalize the orientation, i.e. cast
+    its values from [-1,1] to [0,1]
+    """
+    min_ = typing.cast(Orientation3D, tuple([-1.0] * 9))
+    max_ = typing.cast(Orientation3D, tuple([+1.0] * 9))
```
[the remark this answers is not in the document]
Comment (vincentberenz) [reply hysr/normalize.py:144]: +1

File: hysr/normalize.py [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+    if position_box is not None:
+        return typing.cast(Point3D, normalize(point, position_box[0], position_box[1]))
+    if max_velocity is not None:
+        min_ = (-max_velocity, -max_velocity, -max_velocity)
+        max_ = (+max_velocity, +max_velocity, +max_velocity)
+        return typing.cast(Point3D, normalize(point, min_, max_))
+    # will never happen. The last "if" is there
+    # only to help mypy (which complains about max_velocity
+    # being possibly None)
+    return (0.0, 0.0, 0.0)
```
[the remark this answers is not in the document]
Comment (vincentberenz) [reply hysr/normalize.py:135]: +1

File: hysr/observations.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    # - MainSimState.contact (context.ContactInformation)-> cast to 1 (contact)
+    #                                          or 0 (no contact)
+    # - MainSimState.racket_cartesian: tuple of tuple -> cast to a tuple
+    #   (e.g. ((1,1),(2,2))->(1,1,2,2)
+    # - iteration and time_stamp: no included in _pack (ValueError raised)
+    #   if in attributes
+    values: typing.List[float] = []
+    for attr in attributes:
+        if attr == "contact":
+            if getattr(getattr(state, attr), "contact_occured"):
```
[the remark this answers is not in the document]
Comment (vincentberenz) [reply hysr/observations.py:37]: +1

File: hysr/observations.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    """
+
+    r: typing.List[Observation] = []
+
+    for main_attr, sub_attrs in config:
+        if main_attr == "pressure_robot":
+            if min_pressures is None or max_pressures is None:
+                raise ValueError(
+                    "pack: min and max pressures should not"
+                    "be none when packing an instance of PressureRobotState"
```
[the remark this answers is not in the document]
Comment (vincentberenz) [reply hysr/observations.py:149]: +1

**Pull request merged**